rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+  A ROS service client.
+  Sends requests to services and receives responses.
+:Request:
+  A ROS service request.
+:Response:
+  A ROS service response.
+
+
+Motivation
+==========
+
+The primary motivation for this proposal is to make it easier for users to externally validate that services are operating as expected.
+Drawing an analogy to ROS topics, there exists tools and libraries for "echoing" and recording messages sent over a topic, and we would like to see the same kind of capabilities for services.
+Specifically, we want the capability to introspect requests and responses that are sent to and from services and clients.
+Being able to remotely monitor services allows users to more effectively troubleshoot issues in a ROS system.
+For example, we could verify requests are being received by a service by employing a command-line tool during runtime.
+Or, we could post-process recorded requests and responses to validate their content.
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Request event messages shall be published to the hidden topic ``/SERVICE_NAME/_request_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Similarly, response event messages shall be published to the hidden topic ``/SERVICE_NAME/_response_event``.
+Note that this implies that services must have unique names.
+
+By publishing service event messages to predetermined topics, tools and libraries are able to subscribe to these topics to inspect the flow of data between services.
+
+Service Event Definitions
+-------------------------
+
+For each service definition, ``my/srv/Foo``, we define two new ROS message types using the ROS IDL specification [2]_:
+
+:my/msg/Foo_RequestEvent:
```

Comment:
I would think from a watchdog perspective, having both the request and response events on the same topic would be more convenient. I'm trying to think about how I would use this information in my robots and having 2 separate callbacks for the requests and responses to compare adds more complexity that I don't think is necessary since both messages could contain a field indicating a request vs response.

Reply:
Having both the request and response events on the same topic would add complexity on the user side as the `ClientEvent` or `ServiceEvent` will have to support carrying both request and response content and it would be up to the user to process that. Like you said, it can also be argued that more or less the same applies to the currently proposed implementation with having to monitor two topics to "watchdog" a request/response pair, so it's really a matter of picking the least inconvenient option.

Or we could just mangle everything together into a single slightly more indigestible topic and leave it at that.

Reply:
If we go with a single topic, than we should define a single message type for that topic. As an exercise, here's what I imagine the message definition will look like:

```
# Possible event types
uint8 REQUEST_SENT      = 0
uint8 REQUEST_RECEIVED  = 1
uint8 RESPONSE_SENT     = 2
uint8 RESPONSE_RECEIVED = 3

# The kind of event this message represents
uint8 event_type

# Timestamp for when the event occurred (sent or received time)
builtin_interfaces/msg/Time stamp

# Unique identifier for the client that sent the service request
# Note, this is only unique for the current session
unique_identifier_msgs/msg/UUID client_id

# Sequence number for the request
# Combined with the client ID, this creates a unique ID for the service transaction
int64 sequence_number

# The actual request content sent or received
# Setting this field is optional
my/srv/Foo_Request[<=1] request

# The actual response content sent or received
# Setting this field is optional
my/srv/Foo_Response[<=1] response
```

In the above example, I think only one of the `request` and `response` members would ever be set, to avoid unnecessarily bloating message sizes.

I don't feel strongly one way or the other, though two separate message types seems more natural to me. It also makes it easy to record only requests, for example, though maybe that's not so interesting.

Reply:
In ROS1, we record a lot of bagfiles of various tasks that a robot does using a ROS1 action - and when visualizing a bagfile (say with rqt_bag) - there is a very nice visual indicator of when a task started and ended because the goal/response are on different topics. That is about the only thing I think you would lose by putting everything in one topic.

Reply:
At least in the case of actions, you would still have different topics for the goal service event and result service event.

Reply:
Practically, how would the ROS2 tooling be implemented if the request and response were to be stitched together? Presumably it would need to be implemented in either the server or the client c++/python API, and would that not create problems for a lossy-network context? For example, if the stitching is implemented in the client, and the response is lost, we cannot capture any information about whether the response was ever sent because the client does not have access to that information. It's better if both client and server publish their locally-observed events.

Also, for deterministic replay of client/server interactions, having a time stamp for each ingress/egress event is extremely helpful. This is regardless of whether or not we stitch the request and response.

With this in mind, I'm leaving heavily towards to the per-event design, with the topic being the mechanism to bundle interactions on a specific service, and the sequence_number being a way to link events related to one call.

Reply:
I agree that it is better if the client and server independently publish their own events. I think we can still have separate events with a single message type per service. The request and response members of the message can be mutually exclusive (ie. only one of them will ever be set, depending on the event type).

Reply:
I've updated the REP to define a single message type and one topic per service.